crates/core/src/delta_datafusion/cdf/scan_utils.rs

```diff
@@ -0,0 +1,99 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use chrono::TimeZone;
+use datafusion::datasource::listing::PartitionedFile;
+use datafusion_common::ScalarValue;
+use object_store::path::Path;
+use object_store::ObjectMeta;
+use serde_json::Value;
+
+use crate::delta_datafusion::cdf::scan::CHANGE_TYPE_COL;
+use crate::delta_datafusion::cdf::{CdcDataSpec, FileAction};
+use crate::delta_datafusion::{get_null_of_arrow_type, to_correct_scalar_value};
+use crate::DeltaResult;
+
+pub fn map_action_to_scalar<F: FileAction>(
+    action: &F,
+    part: &str,
+    schema: SchemaRef,
+) -> ScalarValue {
+    action
+        .partition_values()
+        .get(part)
+        .map(|val| {
+            schema
+                .field_with_name(part)
+                .map(|field| match val {
+                    Some(value) => to_correct_scalar_value(
+                        &Value::String(value.to_string()),
+                        field.data_type(),
+                    )
+                    .unwrap_or(Some(ScalarValue::Null))
+                    .unwrap_or(ScalarValue::Null),
+                    None => get_null_of_arrow_type(field.data_type()).unwrap_or(ScalarValue::Null),
+                })
+                .unwrap_or(ScalarValue::Null)
+        })
+        .unwrap_or(ScalarValue::Null)
+}
+
+pub fn create_spec_partition_values<F: FileAction>(
```

Comment:
should have some docs for a `pub` func

Reply:
These are pub in their module but are actually only exported as `pub(crate)` outside the module. There actually is a deny lint set for publicly reachable things that don't have docs